packages/babel-helpers/src/helpers/wrapAsyncGenerator.ts

```diff
@@ -0,0 +1,162 @@
+
+  constructor(gen: Generator<T, TReturn, TNext>);
+
+  next(...args: [] | [TNext]): Promise<IteratorResult<T, TReturn>>;
+  return(
+    value: TReturn | PromiseLike<TReturn>,
+  ): Promise<IteratorResult<T, TReturn>>;
+  throw(e: any): Promise<IteratorResult<T, TReturn>>;
+  [Symbol.asyncIterator](): AsyncGenerator<T, TReturn, TNext>;
+}
+
+interface AsyncGeneratorRequest<T = unknown, TReturn = any, TNext = unknown> {
+  key: AsyncIteratorMethod;
+  arg: IteratorResult<T>;
+  resolve: (value: IteratorResult<T, TReturn>) => void;
+  reject: (error: any) => void;
+  next: AsyncGeneratorRequest<T, TReturn, TNext> | null;
+}
+
+function AsyncGenerator<T = unknown, TReturn = any, TNext = unknown>(
+  this: AsyncGenerator<T, TReturn, TNext>,
+  gen: Generator<T, TReturn, TNext>,
+) {
+  var front: AsyncGeneratorRequest<T, TReturn, TNext> | null,
+    back: AsyncGeneratorRequest<T, TReturn, TNext> | null;
+
+  function send(key: AsyncIteratorMethod, arg: IteratorResult<T>) {
+    return new Promise<IteratorResult<T, TReturn>>(function (resolve, reject) {
+      var request: AsyncGeneratorRequest<T, TReturn, TNext> = {
+        key: key,
+        arg: arg,
+        resolve: resolve,
+        reject: reject,
+        next: null,
+      };
+
+      if (back) {
+        back = back.next = request;
+      } else {
+        front = back = request;
+        resume(key, arg);
+      }
+    });
+  }
+
+  function resume(key: AsyncIteratorMethod, arg: IteratorResult<T, TReturn>) {
+    try {
+      var result = gen[key](arg);
+      var value = result.value;
+      var overloaded = value instanceof OverloadYield;
+
+      Promise.resolve(
+        overloaded ? (value as OverloadYield<T | TReturn>).v : value,
+      ).then(
+        function (arg: any) {
+          if (overloaded) {
+            // Overloaded yield requires calling into the generator twice:
+            //  - first we get the iterator result wrapped in a promise
+            //    (the gen[key](arg) call above)
+            //  - then we await it (the Promise.resolve call above)
+            //  - then we give the result back to the iterator, so that it can:
+            //    * if it was an await, use its result
+            //    * if it was a yield*, possibly return the `done: true` signal
+            //      so that yield* knows that the iterator is finished.
+            //      This needs to happen in the second call, because in the
+            //      first one `done: true` was hidden in the promise and thus
+            //      not visible to the (sync) yield*.
+            //      The other part of this implementation is in asyncGeneratorDelegate.
+            var nextKey: "return" | "next" =
+              key === "return" ? "return" : "next";
+            if (
+              !(value as OverloadYield<IteratorReturnResult<T>>).k ||
+              arg.done
+            ) {
+              // await or end of yield*
+              resume(nextKey, arg);
+              return;
```

Comment:
```suggestion
              return resume(nextKey, arg);
```

does TS complain if you do this? It leads to slightly smaller minified code.

Reply:
@nicolo-ribaudo It is the ESLint rule [`no-confusing-void-expression`](https://typescript-eslint.io/rules/no-confusing-void-expression/) that complains about this. Let me see if I can workaround this using explicit types.

Reply:
We can just use `eslint-disable-next-line` here.

Reply:
Done in [`fa8338f` (#16510)](https://github.com/babel/babel/pull/16510/commits/fa8338fa8d1ac047d1f3427017a3e9402b382818).